### PR TITLE
Add tests generation for all combinations

### PR DESCRIPTION
**Pull request by Warchant**
- fix https://github.com/hyperledger/iroha-ed25519/issues/6
- add automatic test generation: since we use multiple EDIMPL, HASH, RANDOM implementations, which can be selected at link-time, it is hard to test everything on CI. With automatic test generation it is a little bit easier - all combinations are tested.
- setup travis CI to test library on Linux and Mac, with both clang and gcc

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,21 +119,69 @@ ENUM(CMAKE_BUILD_TYPE "${CMAKE_BUILD_TYPE}" "cmake build type"
...
+    ${RANDOM}
+  VERSION
+    VERSION-${SOVERSION}-${EDIMPL}-${HASH}-${RANDOM}-${BUILD}-${CMAKE_BUILD_TYPE}
+  )
+set(_export_file ${CMAKE_BINARY_DIR}/ed25519_export.h)
+generate_export_header(ed25519
+  EXPORT_FILE_NAME ${_export_file}
+  )
+install(
+  FILES ${_export_file}
```
Comment (l4l) [inline CMakeLists.txt]: That won't work for projects that use it as an external one (e.g iroha)

File: .travis.yml [hunk 2]
```diff
@@ -26,27 +25,35 @@ matrix:
...
-  - mkdir build
-  - cd build
-  - export CC=$COMPILERCC; export CXX=$COMPILERCXX
-  - cmake .. -DEDIMPL=ref10 -DHASH=sha3_brainhub -DRANDOM=dev_urandom
+  - export CC=$_CC
+  - export CXX=$_CXX
+  # build and test ed25519
+  - mkdir build && cd build
+  - cmake .. -DTESTING=ON
+  - make -j2 -k
```
[the remark this answers is not in the document]
Comment (Warchant) [reply .travis.yml]: -k continues build even if build errors are present. I want to know which combinations have failed, e.g. all of them.

File: .travis.yml [hunk 3]
```diff
@@ -26,27 +25,35 @@ matrix:
...
+  - ctest
+  - sudo make install
+  # build and execute example
+  - cd ../example/sign-verify
+  - mkdir build && cd build
+  - cmake .. -DCMAKE_BUILD_TYPE=Release
```
[the remark this answers is not in the document]
Comment (Warchant) [reply .travis.yml]: When you write `make install`, special `ed25519-Config.cmake` file is installed. When you write `find_package(ed25519)`, cmake locates this file and includes.

This example is used to test the behavior of this file and targets defined inside.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,51 +1,74 @@
...
+enable_language(ASM)
 
 set(CMAKE_POSITION_INDEPENDENT_CODE    TRUE)
+set(CMAKE_C_VISIBILITY_PRESET          hidden)
+set(CMAKE_CXX_VISIBILITY_PRESET        hidden)
+set(CMAKE_VISIBILITY_INLINES_HIDDEN    TRUE)
 set(CMAKE_EXPORT_COMPILE_COMMANDS      ON)
 set(CMAKE_MODULE_PATH                  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules")
-set(EXECUTABLE_OUTPUT_PATH             "${PROJECT_BINARY_DIR}/bin")
-set(LIBRARY_OUTPUT_PATH                "${PROJECT_BINARY_DIR}")
```
[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: `ed25519_*` functions contain these vars as properties.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,51 +1,74 @@
...
 set(CMAKE_CXX_EXTENSIONS               OFF)
-set(CMAKE_CXX_FLAGS_RELEASE            "${CMAKE_CXX_FLAGS_RELEASE} -O3 -DNDEBUG")
-set(CMAKE_CXX_FLAGS_DEBUG              "${CMAKE_CXX_FLAGS_DEBUG} -g -O0 -fdiagnostics-color")
 
-add_compile_options(-Wall)
-add_compile_options(-Wextra)
-add_compile_options(-Wpedantic)
+#add_compile_options(-Wall)
+#add_compile_options(-Wextra)
+#add_compile_options(-Wpedantic)
```
[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: Not all compilers support these flags. (I am talking about MSVC)

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,51 +1,74 @@
...
+include(GenerateExportHeader)
 include(cmake/dependencies.cmake)
 include(cmake/functions.cmake)
-enable_language(ASM)
+include(cmake/ed25519_init.cmake)
+include(cmake/ed25519_export.cmake)
+include(cmake/ed25519_target_link_libraries.cmake)
+include(cmake/ed25519_add_library.cmake)
+include(cmake/ed25519_merge_libraries.cmake)
+include(cmake/ed25519_add_test.cmake)
```
[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: It is hard to maintain big file and easy to do separation like this.

File: CMakeLists.txt (hunk 1 above)
Comment (Warchant) [reply CMakeLists.txt]: Why?

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: Next patches with Windows fixes will contain these flags.

File: CMakeLists.txt (hunk 1 above)
Comment (l4l) [reply CMakeLists.txt]: Try to make a clean build of iroha, it doesn't work that for me

**Pull request merged**